eclair-core/src/main/scala/fr/acinq/eclair/wire/LightningMessageTypes.scala

```diff
@@ -229,69 +229,150 @@ case class PerHopPayload(shortChannelId: ShortChannelId,
+
+  def includeUpdate2(flag: Long) = (flag & INCLUDE_CHANNEL_UPDATE_2) != 0
+}
+
+/**
+  * Optional query flag that is appended to QueryChannelRange
+  * @param flag bit 1 set means I want timestamps, bit 2 set means I want checksums
+  */
+case class QueryChannelRangeExtension(flag: Long) extends Tlv {
+  val wantTimestamps = QueryChannelRangeExtension.wantTimestamps(flag)
+
+  val wantChecksums = QueryChannelRangeExtension.wantChecksums(flag)
+}
+
+case object QueryChannelRangeExtension {
+  val WANT_TIMESTAMPS: Long = 1
+  val WANT_CHECKSUMS: Long = 2
+  val WANT_ALL: Long = (WANT_TIMESTAMPS | WANT_CHECKSUMS)
+
+  def wantTimestamps(flag: Long) = (flag & WANT_TIMESTAMPS) != 0
+
+  def wantChecksums(flag: Long) = (flag & WANT_CHECKSUMS) != 0
+}
+
+
+/**
+  *
+  * @param timestamp1 timestamp for node 1, or 0
+  * @param timestamp2 timestamp for node 2, or 0
+  */
+case class Timestamps(timestamp1: Long, timestamp2: Long)
+
+/**
+  * Optional timestamps TLV that can be appended to ReplyChannelRange
+  * @param encoding same convention as for short channel ids
+  * @param timestamps
+  */
+case class EncodedTimestamps(encoding: EncodingType, timestamps: List[Timestamps]) extends Tlv
+
+/**
+  *
+  * @param checksum1 checksum for node 1, or 0
+  * @param checksum2 checksum for node 2, or 0
+  */
+case class Checksums(checksum1: Long, checksum2: Long)
+
+/**
+  * Optional checksums TLV that can be appended to ReplyChannelRange
+  * There is no leading encoding byte, as compression would ne be very effective
+  * @param checksums
+  */
+case class EncodedChecksums(checksums: List[Checksums]) extends Tlv
+
+// @formatter:on
 sealed trait EncodingType
 object EncodingType {
   case object UNCOMPRESSED extends EncodingType
   case object COMPRESSED_ZLIB extends EncodingType
 }
 // @formatter:on
 
-case object QueryFlagTypes {
-  val INCLUDE_CHANNEL_ANNOUNCEMENT: Byte = 1
-  val INCLUDE_CHANNEL_UPDATE_1: Byte = 2
-  val INCLUDE_CHANNEL_UPDATE_2: Byte = 4
-  val INCLUDE_ALL: Byte = (INCLUDE_CHANNEL_ANNOUNCEMENT | INCLUDE_CHANNEL_UPDATE_1 | INCLUDE_CHANNEL_UPDATE_2).toByte
-
-  def includeAnnouncement(flag: Byte) = (flag & QueryFlagTypes.INCLUDE_CHANNEL_ANNOUNCEMENT) != 0
-
-  def includeUpdate1(flag: Byte) = (flag & QueryFlagTypes.INCLUDE_CHANNEL_UPDATE_1) != 0
-
-  def includeUpdate2(flag: Byte) = (flag & QueryFlagTypes.INCLUDE_CHANNEL_UPDATE_2) != 0
-}
 
 case class EncodedShortChannelIds(encoding: EncodingType,
                                   array: List[ShortChannelId])
 
-case class EncodedQueryFlags(encoding: EncodingType,
-                             array: List[Byte])
 
 case class QueryShortChannelIds(chainHash: ByteVector32,
+                                shortChannelIds: EncodedShortChannelIds,
+                                extensions: TlvStream[Tlv]) extends RoutingMessage with HasChainHash {
+  val queryFlags: Option[EncodedQueryFlags] = extensions.records collectFirst { case flags: EncodedQueryFlags => flags }
```

Comment:
Should we make a generic implicit `.getRecord[T]` available for all TLV, as we will always use this pattern? cc @t-bast

Reply:
That's a good idea!

Reply:
Hum looks like we can't do that without reflection because of type erasure...
I'll let the Scala guru try to find a way to do this but my naive attempt didn't work, Scala generics aren't C++ templates ;)

Reply:
We can simply do:
```scala
case class TlvStream[T <: Tlv](records: Traversable[T], unknown: Traversable[GenericTlv] = Nil) {
  def get[R <: T](implicit m: Manifest[R]): Option[R] = records.collectFirst { case r: R => r }
}
```

And then:
```scala
test("extract record by type") {
  val stream = TlvStream[TestTlv](Seq(TestType254(42), TestType1(42)), Seq(GenericTlv(13, hex"2a"), GenericTlv(11, hex"2b")))
  assert(stream.get[TestType1].contains(TestType1(42)))
  assert(stream.get[TestType2].isEmpty)
  assert(stream.get[OtherType1].isEmpty) // <- doesn't compile
}
```

😎

Reply:
Yes but if we use `Manifest` this won't work an Android right?

Reply:
I'm not sure, let's ask @sstone

Reply:
Not sure how yet but it does seem to work on Android (with a real device) I'll update this discussion asap

Reply:
Implemented with:

``` scala
case class TlvStream[T <: Tlv](records: Traversable[T], unknown: Traversable[GenericTlv] = Nil) {
  def get[R <: T : ClassTag]: Option[R] = records.collectFirst { case r: R => r }
}
```